pkg/pool/pool.go

```diff
@@ -108,28 +101,67 @@ func newPool(ctx context.Context, options *BuilderOptions) (Pool, error) {
 	ownerID := owner.NewIDFromNeo3Wallet(wallet)
 
 	pool := &pool{sampler: sampler, owner: ownerID, clientPacks: clientPacks}
-	go func() {
-		ticker := time.NewTimer(options.ClientRebalanceInterval)
-		for range ticker.C {
-			ok := true
-			for i, clientPack := range pool.clientPacks {
-				func() {
-					tctx, c := context.WithTimeout(ctx, options.NodeRequestTimeout)
-					defer c()
-					if _, err := clientPack.client.EndpointInfo(tctx); err != nil {
-						ok = false
-					}
-					pool.lock.Lock()
-					pool.clientPacks[i].healthy = ok
-					pool.lock.Unlock()
-				}()
-			}
-			ticker.Reset(options.ClientRebalanceInterval)
-		}
-	}()
+	go startRebalance(ctx, pool, options)
 	return pool, nil
 }
 
+func startRebalance(ctx context.Context, p *pool, options *BuilderOptions) {
+	ticker := time.NewTimer(options.ClientRebalanceInterval)
+	newWeights := make([]float64, len(options.weights))
+
+	for range ticker.C {
+		healthyChanged := false
+		wg := sync.WaitGroup{}
+		for i, cPack := range p.clientPacks {
+			go func(i int, netmap client.Netmap) {
+				wg.Add(1)
```

Comment:
That's racy, it should be done one line above. Think of the main thread spawning goroutines, but none of them starting to execute before `wg.Wait()`.